demo_modules/wind/forecast_grid.js

```diff
@@ -0,0 +1,139 @@
+
+const floorMod = util.floorMod;
+const isValue = util.isValue;
+
+
+// Currently unused.
+function bilinearInterpolateScalan(x, y, g00, g10, g01, g11) {
+	var rx = (1 - x);
+	var ry = (1 - y);
+	return g00 * rx * ry + g10 * x * ry + g01 * rx * y + g11 * x * y;
+}
+
+function bilinearInterpolateVector(x, y, g00, g10, g01, g11) {
+	var rx = (1 - x);
+	var ry = (1 - y);
+	var a = rx * ry,  b = x * ry,  c = rx * y,  d = x * y;
+	var u = g00[0] * a + g10[0] * b + g01[0] * c + g11[0] * d;
+	var v = g00[1] * a + g10[1] * b + g01[1] * c + g11[1] * d;
+	return [u, v, Math.sqrt(u * u + v * v)];
+}
+
+class Forecast {
+  constructor(gfsJson) {
+    this.uData = gfsJson[0].data;
+    this.vData = gfsJson[1].data;
+    this.header = gfsJson[0].header;
+	}
+
+  get(i) {
+    return [this.uData[i], this.vData[i]];
+  }
+}
+
+class ForecastGrid {
+	constructor(forecastJson) {
+    const forecast = new Forecast(forecastJson);
+		const header = forecast.header;
+
+		// the grid's origin (e.g., 0.0E, 90.0N)
+		this.λ0 = header.lo1;
+    this.φ0 = header.la1;
+
+    // distance between grid points (e.g., 2.5 deg lon, 2.5 deg lat)
+		this.Δλ = header.dx;
+    this.Δφ = header.dy;
+
+    // number of grid points W-E and N-S (e.g., 144 x 73)
+		this.ni = header.nx;
+    this.nj = header.ny;
+
+		// Scan mode 0 assumed. Longitude increases from λ0, and latitude decreases from φ0.
+		// http://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_table3-4.shtml
+		const grid = [];
+    var p = 0;
+		var isContinuous = Math.floor(this.ni * this.Δλ) >= 360;
+		for (var j = 0; j < this.nj; j++) {
+			var row = [];
+			for (var i = 0; i < this.ni; i++, p++) {
+				row[i] = forecast.get(p);
+			}
+			if (isContinuous) {
+				// For wrapped grids, duplicate first column as last column to simplify interpolation logic
+				row.push(row[0]);
+			}
+			grid[j] = row;
+		}
+
+    this.grid = grid;
+		this.date = new Date(header.refTime);
+		this.date.setHours(this.date.getHours() + header.forecastTime);
+  }
+
+	interpolate(λ, φ) {
+    // calculate longitude index in wrapped range [0, 360)
+		var i = floorMod(λ - this.λ0, 360) / this.Δλ;
+
+    // calculate latitude index in direction +90 to -90
+		var j = (this.φ0 - φ) / this.Δφ;
+
+		//         1      2           After converting λ and φ to fractional grid indexes i and j, we find the
+		//        fi  i   ci          four points "G" that enclose point (i, j). These points are at the four
+		//         | =1.4 |           corners specified by the floor and ceiling of i and j. For example, given
+		//      ---G--|---G--- fj 8   i = 1.4 and j = 8.3, the four surrounding grid points are (1, 8), (2, 8),
+		//    j ___|_ .   |           (1, 9) and (2, 9).
+		//  =8.3   |      |
+		//      ---G------G--- cj 9   Note that for wrapped grids, the first column is duplicated as the last
+		//         |      |           column, so the index ci can be used without taking a modulo.
+
+		var fi = Math.floor(i), ci = fi + 1;
```

Comment:
Probably move this to a utility method, as bilinearinterpolation is a useful construct by itself.